### PR TITLE
refactor(core): refactor interaction guard

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

Refactor Interaction guard.

This PR includes following updates:
1. make the `event` payload field rerquired
2. Split and detailed guard the interaction payloads by the event type
  - e.g. forgot-password should only accept password in its profile payload, and passcode identifiers in its identifier payload
3.  Do not guard the SIE settings if the event is forgot-password

## Testing

UT case updated

### CODE REVIEW

File: packages/core/src/routes/interaction/middleware/koa-session-sign-in-experience-guard.ts [hunk 1]
```diff
@@ -20,13 +20,15 @@ export default function koaSessionSignInExperienceGuard<
     const interaction = await provider.interactionDetails(ctx.req, ctx.res);
     const { event, identifier, profile } = ctx.interactionPayload;
 
+    if (event === 'forgot-password') {
```
Comment (darcyYe) [inline packages/core/src/routes/interaction/middleware/koa-session-sign-in-experience-guard.ts:23]: as this is `koaSessionSignInExpecrienceGuard`, should we also consider 'register' event as not applicable?

File: packages/core/src/routes/interaction/types/guard.ts [hunk 2]
```diff
@@ -6,16 +6,43 @@ import type {
...
+const forgotPasswordInteractionPayloadGuard = z.object({
+  event: z.literal('forgot-password'),
+  identifier: z.union([emailPasscodePayloadGuard, phonePasscodePayloadGuard]).optional(),
+  profile: profileGuard.pick({ password: true }).optional(),
+});
+
+const registerInteractionPayloadGuard = z.object({
+  event: z.literal('register'),
+  identifier: z
+    .union([emailPasscodePayloadGuard, phonePasscodePayloadGuard, socialConnectorPayloadGuard])
```
Comment (darcyYe) [inline packages/core/src/routes/interaction/types/guard.ts:29]: password identifiers can not be used to register?

File: packages/core/src/routes/interaction/middleware/koa-session-sign-in-experience-guard.ts (hunk 1 above)
Comment (simeng-li) [reply packages/core/src/routes/interaction/middleware/koa-session-sign-in-experience-guard.ts:23]: nope we guard the register methods as well.

File: packages/core/src/routes/interaction/types/guard.ts (hunk 2 above)
Comment (simeng-li) [reply packages/core/src/routes/interaction/types/guard.ts:29]: email and phone identifier must be passcode verified before registeation. Password is allowed in the profile.

**Review by darcyYe — APPROVED** [review]

**Pull request merged**